cpp/src_prims/selection/knn.h

```diff
@@ -104,93 +170,451 @@ void merge_tables(long n, long k, long nshard, float *distances, long *labels,
    * @param res_I      pointer to device memory for returning k nearest indices
    * @param res_D      pointer to device memory for returning k nearest distances
    * @param k        number of neighbors to query
+   * @param allocator the device memory allocator to use for temporary scratch memory
    * @param s the cuda stream to use
+   * @param internalStreams optional when n_params > 0, the index partitions can be
+   *        queried in parallel using these streams. Note that n_int_streams also
+   *        has to be > 0 for these to be used and their cardinality does not need
+   *        to correspond to n_parts.
+   * @param n_int_streams size of internalStreams. When this is <= 0, only the
+   *        user stream will be used.
+   * @param translations translation ids for indices when index rows represent
+   *        non-contiguous partitions
    */
-template <typename IntType = int>
-void brute_force_knn(float **input, int *sizes, int n_params, IntType D,
-                     float *search_items, IntType n, long *res_I, float *res_D,
-                     IntType k, cudaStream_t s) {
-  std::vector<long> *id_ranges = new std::vector<long>();
+template <typename IntType = int,
+          Distance::DistanceType DistanceType = Distance::EucUnexpandedL2>
+void brute_force_knn(std::vector<float *> &input, std::vector<int> &sizes,
+                     IntType D, float *search_items, IntType n, int64_t *res_I,
+                     float *res_D, IntType k,
+                     std::shared_ptr<deviceAllocator> allocator,
+                     cudaStream_t userStream,
+                     cudaStream_t *internalStreams = nullptr,
+                     int n_int_streams = 0, bool rowMajorIndex = true,
+                     bool rowMajorQuery = true,
+                     std::vector<int64_t> *translations = nullptr) {
+  ASSERT(DistanceType == Distance::EucUnexpandedL2 ||
+           DistanceType == Distance::EucUnexpandedL2Sqrt,
+         "Only EucUnexpandedL2Sqrt and EucUnexpandedL2 metrics are supported "
+         "currently.");
+
+  ASSERT(input.size() == sizes.size(),
+         "input and sizes vectors should be the same size");
+
+  int n_params = input.size();
+
+  std::vector<int64_t> *id_ranges;
+  if (translations == nullptr) {
+    // If we don't have explicit translations
+    // for offsets of the indices, build them
+    // from the local partitions
+    id_ranges = new std::vector<int64_t>();
+    int64_t total_n = 0;
+    for (int i = 0; i < n_params; i++) {
+      if (i < input.size()) {
```

Comment:
do we need this check when the value of `n_params` is basically `input.size()`?

Reply:
This should be `input.size()-1`. It doesn't effect correctness, but it does add an extra element to the array which will never get used. I should just go ahead and remove the `n_params` altogether.